### ISSUE TITLE
email(s) not being sent when test plan or test case change

### ISSUE THREAD

**Issue by bkosanovic** [kiwitcms/Kiwi#3212]
### Description of problem

The problem was first reported at the [stackoverflow here](https://stackoverflow.com/questions/76326570/kiwi-tcms-v12-2-does-not-send-email-when-test-plan-or-test-cases-change).

After extensive debugging, testing, etc., it appears that the email messages are most likely not even attempted to be sent when test plan is updated or test case is updated.

Please note that test message and password change message are being sent and received. However, when a test plan is changed or test case, the message(s) are not being sent although notifications were properly configured for test plan and test cases.

### Version or commit hash (if applicable)

v12.1, v12.2, v12.3 were tested and all behave the same way

### Steps to Reproduce

1. Verify that test message is going
2. Change test plan description or test case description making sure notifications are enabled.
3. Save the changed test plan or test case

### Actual results

No messages are being sent (verified on SMTP server that none of the message were received nor bounced nor rejected during the time when the changes in Kiwi TCMS were saved.

### Expected results

email messages should be sent to appropriate enabled recipients when test plan or test cases were saved.

### Additional info (Python traceback, logs, etc.)

n/a

**Comment by somenewacc** [comment]
Emails are sent in a separate thread.
Actually, uWSGI does not goes well with threads.
There are some possible solutions such as enabling threads in config (https://uwsgi-docs.readthedocs.io/en/latest/Options.html#enable-threads).

On our project, we also use threads for other tasks in Kiwi, so we switched back to Apache, because we could not get uWSGI to work properly

**Comment by bkosanovic** [comment]
Well, we did not choose to use uWSGI. We just installed Kiwi following your instructions on ubuntu VM that is only used for Kiwi and nothing else. You should not be providing configurations that do not work. We are not looking for "possible" solutions, but those that are tested and that work. It would be nice if you would provide software that works. After so many releases, one would expect for it to be stable without so many bugs being introduced or discovered within basic functionality as email notifications.

Putting the obvious lack of proper testing and complete / verified installation instructions aside, could you please provide instructions on how to properly install Kiwi so that the email notifications would work correctly. (you may want to consider implementing this on your kiwitcms.org test site as well since it fails there as well).

Honestly, we are at the tipping point where we may just drop Kiwi TCMS from consideration given the number of issues we are finding so easily and so frequently. It is very nice on the surface and convenient (when it works), but given the issues we are finding we do not see it much better compared to using Excel spreadsheet for test cases and Confluence page for test plan and test case descriptions.

**Comment by somenewacc** [comment]
@bkosanovic 
I'm just a user, just like you. Not a main contributor :)

Kiwi actually has some flaws (sometimes kinda major), despite its advanced age. Perhaps the problem is the lack of popularity of the product (and as a result, lack of support). That's only my opinion.

But I still like it, even though I had to patch it enough.

**Comment by bkosanovic** [comment]
Ok. Well, thank you for offering help! :)

I like Kiwi too. But it is a bit frustrating to keep finding basic functionality broken. We are trying to deploy it, but we can't do it as long as we are finding this type of flaws. The lack of popularity and lack of support are like the chicken and egg in many ways. :) Although I think it is more about lack of basic functionality working as advertised. :)

For the things you suggested, would I need to rebuild the docker image or it can be done in a less intrusive way?

**Comment by somenewacc** [comment]
If you use docker-compose, you can try this:
1. Create file with uWSGI config (use https://github.com/kiwitcms/Kiwi/blob/master/etc/uwsgi.conf as example). Add line:
`enable-threads = true` (after `master = true` for example)
2. Make changes to your docker-compose file in web->volumes:
```
volumes:
    - <PATH-TO-NEW-CONFIG>/uwsgi.conf:/Kiwi/etc/uwsgi.conf
    - uploads:/Kiwi/uploads:Z
```
3. Run container: `docker-compose up -d`

I didn't personally tested emails with this solution but you can try.

**Comment by somenewacc** [comment]
@bkosanovic 
Actually, I see that emails are working on [public.tenant.kiwitcms.org](https://public.tenant.kiwitcms.org/):
![image](https://github.com/kiwitcms/Kiwi/assets/62895232/0db75a4e-f695-4109-a67f-4d42c79f8fa3)

I tested locally and I can see emails. So that's can be not about uWSGI.

You can add to CC list in Test Case edit view and check it:
![image](https://github.com/kiwitcms/Kiwi/assets/62895232/6002bb7b-2399-4aeb-93b1-8aebef2fd9ff)

BTW, you can check if emailing is working via `manage.py`:
https://kiwitcms.readthedocs.io/en/latest/configuration.html#testing-and-debugging-e-mail-configuration

**Comment by bkosanovic** [comment]
I tested emails to be working using `docker exec -it kiwi_web /Kiwi/manage.py sendtestemail someone@example.com` 
That works OK. I just could not get it to work when I change the test plan and test cases.

On the kiwitcms.org I tried to change test plan and it did not send message to me (I was the author) and author was checked on to receive the notifications. I will test it with test cases on *.org. On our installation it did not work. It did not send messages to "author" or other roles and it did not send messages to CC list.

I'll post here my findings with your suggested change and online test.

**Comment by bkosanovic** [comment]
I tried again kiwitcms.org site and if did not work again. I changed the test case and I did not get any email. I tried CC as well using 3 different e-mails 2 gmail and one non-gmail. I checked the SPAM folders. Nothing! :(

Not sure why it would work for you and not for me :) After all it is the online test site which I have nothing to do with.

I used it from here: https://public.tenant.kiwitcms.org/ It seems to be running 12.3-Enterprise version

**Comment by somenewacc** [comment]
Could you please give me a link to tested case?

**Comment by bkosanovic** [comment]
Test case: https://public.tenant.kiwitcms.org/case/17913/
Test plan: https://public.tenant.kiwitcms.org/plan/7181/test-plan
Test run: https://public.tenant.kiwitcms.org/runs/6207/

**Comment by bkosanovic** [comment]
Just got this e-mail:
![image](https://github.com/kiwitcms/Kiwi/assets/11080474/f55c9099-1814-4ce1-8d2b-6119e0a64c03)

How come it likes you and it does not like me? What exactly did you do for this to happen?!?

**Comment by bkosanovic** [comment]
Just got another collection of emails for changes you did. Date changes, etc. So far I am not getting emails when I change things.

**Comment by somenewacc** [comment]
Actually IDK. I just changed some things and added you to notify list in test run:
![image](https://github.com/kiwitcms/Kiwi/assets/62895232/10e2a84f-ad22-4a0a-94cc-78a6208c572a)

**Comment by bkosanovic** [comment]
Let me try again (although I do not see why would notify list in test run have anything to do with separately updating test plan, etc. which already has notification by roles, etc.

**Comment by bkosanovic** [comment]
No luck. When I change something in test plan, test cases, etc. I do not get any notifications. When you do it I get emails. I could try on our own installation to get someone else to do changes and see if I would get emails. That is the only remaining possibility. Perhaps it is not sending emails to person who actually performs the change. I somehow doubt that is the case, but who knows.

**Comment by somenewacc** [comment]
> Perhaps it is not sending emails to person who actually performs the change

That can be the case actually :)
![image](https://github.com/kiwitcms/Kiwi/assets/62895232/bcb3edc8-cb0d-4082-8958-354c8dcfca84)

**Comment by atodorov** [comment]
> Just got this e-mail: ![image](https://user-images.githubusercontent.com/11080474/241037501-f55c9099-1814-4ce1-8d2b-6119e0a64c03.png)
> 
> How come it likes you and it does not like me? What exactly did you do for this to happen?!?

@bkosanovic so you have finally established that email notifications do work, even after being very quick to badmouth the project and its entire community of contributors. That's not how open source works!

What you are looking for is the functionality of "email me when I make a change" . That's been disabled long ago, see https://github.com/kiwitcms/Kiwi/blob/master/tcms/testplans/helpers/email.py#LL28C1-L28C1 and https://github.com/kiwitcms/Kiwi/issues/216 and the UI will be made more clear in #3214. 

If you still prefer to receive email notifications even when making changes yourself, then that's a whole different story. Feel free to open a feature request for consideration.

**Comment by bkosanovic** [comment]
I'll do additional testing regarding this for sure. I was expecting to see emails being sent to the roles that are selected to be sent notifications. I could not establish that it was happening. I was additionally prevented in making sense out of it since with 12.2 release for some reason it became impossible to change email address of accounts. So, I could not change email address of some other accounts to test all this. We also do not want to allow registrations through e-mail. So, there was no way to create account with some email address after I upgraded to 12.2 (due to other bug fixes that were required).

Finally, there was nothing in the documentation to indicate notifications would not be sent to those who do the change and there was no reason to assume otherwise. Also, I always suspect that I am doing something wrong and only after considerable time trying to get this to work I decided to ask for help. I was changing versions of Kiwi, I even rebuilt the docker image from scratch (just in case). Bottom line, it took me long time before I decided to communicate about this.

I still do not know why is that the notifications were not sent to additional two different emails that I specified in CC list. I can only guess it is because there was no account associated with those emails. But that is a guess given that in documentations there is nothing that would state it is not allowed.

I will close this issue after I test it additionally on version 12.1 and confirm this is working since on 12.2+ I cannot change email addresses anyways. You should consider allowing admins to change email addresses, etc. Otherwise, we will not be able to use Kiwi in our organization.

**Comment by bkosanovic** [comment]
I just tried it again with V12.1. I created a dummy user with email that is real email but not my email. I tested that email to be working as before using senttestemail command of manage.py. The test email was received. I changed the test plan for which I'm the author using the dummy acount and I did not get email notification that dummy user changed the test plan.

Both emails are within the same domain of our organization.

Go figure. I have no clue why this is still not working even when the change is done by someone who is not the author of the test plan. I'll do additional testing again in this manner for the test cases, etc. and will report back here.

**Comment by bkosanovic** [comment]
Still does not work for test cases and test runs. However, this time I see that some of the messages may have arrived to the SMTP server but may have been bounced. Well, that might be because I used in a few cases the addresses that have not been verified and are not part of our domain.

I have some ideas regarding configuration that I may try. I'll make an effort to complete that testing within next few days and report back here. It still puzzles me a lot that:

1) test emails work (and password change emails)
2) I am using now different account with different email to do the changes, but the author's email or the one from CC list is not getting notification

The 1) and 2) together do not make much sense, but I'll keep trying.

**Comment by bkosanovic** [comment]
I found out what was the problem all along. At least I hope. It will need more testing, but I am finally getting emails to be sent when changes to test plan, test cases occur.

I have switched to use our AWS SES which we also use for a few other IT service alerts. However, the instructions you provided in documentation were still not good enough. It took a few hours to find out that I need to do the following:

1) Do not use the ses-smtp service which is the regular one and that gets access key created when you initially create SMTP Credentials within the Amazon SES > SMTP Settings. That access key will not work for Kiwi TCMS (not sure why)
2) Create a secondary access key for the same ses-smtp-user and select to allow on-premises applications to have access when prompted for the type of the key. That will be a key with different type of access.
3) Within the IAM management for the ses-smtp-user.nnnnn (which is the user created by AWS for this purpose) add additional permission policy to allow more than just default AmazonSesSendingAccess policy created in step 1 above at the time you were creating the SES SMTP credentials. For example, add AmazonSESFullAccess permission policy.
4) Make sure you generate new docker image as explained in documentation so you can use Django SES Backend

Now, in Kiwi TCMS do not use the primary key for SES SMTP, but this secondary key which was created for a different type of access. The policies apply to the user not specific keys.

Finally, the local_settings.py which I came up with looked like this:
```
# EMAIL Settings
SERVER_EMAIL = DEFAULT_FROM_EMAIL = 'kiwitcms@example.com'
EMAIL_SUBJECT_PREFIX = '[Kiwi-TCMS] '

# AWS SES HOST
EMAIL_HOST = '<removed>'

EMAIL_BACKEND = 'django_ses.SESBackend'

AWS_SES_ACCESS_KEY_ID = '<secondary-key-ID>'
AWS_SES_SECRET_ACCESS_KEY = '<secondary-key-secret>'

# Enable TLS
EMAIL_USE_TLS = True
```

This would make both email tests to work and when changes to test cases or test plans are made the notifications will be sent as configured.

Two things should be done as a result of all this:

1) Kiwi TCMS documentation should be updated to better explain how to configure email notifications properly (for AWS SES case it can work). However the documentation should state "complete" collection of settings indicating that port should not be used for SES, HOST user and HOST password should not be used. Documentation currently states:

> If you’d like to use an external email service, like Amazon SES you need to configure the following settings too:

That is wrong. It is not "too" but "instead" still providing the EMAIL_HOST but NOT providing EMAIL_PORT. This confusion may be avoided by simply providing complete settings for both cases AWS SES and generic SMTP.

2) The ability to change email addresses should be allowed for admins. Also, the admins should be allowed to manually add users and specify their email address in the process. (otherwise, in organizations like ours Kiwi TCMS will not be possible to be used due to our IT/Cyber policies)

One additional desirable thing to make it easier to use Kiwi TCMS (Out of the Box Experience):

* Consider adding `django_ses` to the standard Kiwi TCMS docker image so users do not have to jump through hoops to customize the docker image to make email notifications work through Amazon SES that many people use these days (it seems to be used by the public demo on kiwitcms.org as well as far as I could tell)

Please consider these suggestions and you may dispose of this issue based on how you choose to address it.
